cpp/include/raft/cluster/detail/kmeans_auto_find_k.cuh

```diff
@@ -0,0 +1,219 @@
+    centroids_const_view =
+      raft::make_device_matrix_view<const value_t, idx_t>(centroids.data_handle(), right, d);
+    centroids_view =
+      raft::make_device_matrix_view<value_t, idx_t>(centroids.data_handle(), right, d);
+
+    cluster_sizes_view = raft::make_device_vector_view<const idx_t, idx_t>(clusterSizes_ptr, right);
+
+    params.n_clusters = right;
+    raft::cluster::detail::kmeans_fit_predict<value_t, idx_t>(handle,
+                                                              params,
+                                                              X,
+                                                              std::nullopt,
+                                                              std::make_optional(centroids_view),
+                                                              labels.view(),
+                                                              residual,
+                                                              n_iter);
+
+    detail::countLabels(
+      handle, labels.data_handle(), clusterSizes.data_handle(), n, right, workspace);
+
+    resultsView[right]           = residual[0];
+    clusterDispertionView[right] = raft::stats::cluster_dispersion(
+      handle, centroids_const_view, cluster_sizes_view, std::nullopt, n);
+
+    tests += 1;
+  }
+
+  objective[0] = (n - left) / (left - 1) * clusterDispertionView[left] / resultsView[left];
+  objective[1] = (n - right) / (right - 1) * clusterDispertionView[right] / resultsView[right];
+  while (left < right - 1) {
+    resultsView[mid] = 1e20;
+    tests            = 0;
+    while (resultsView[mid] > resultsView[left] && tests < 3) {
+      centroids_const_view =
+        raft::make_device_matrix_view<const value_t, idx_t>(centroids.data_handle(), mid, d);
+      centroids_view =
+        raft::make_device_matrix_view<value_t, idx_t>(centroids.data_handle(), mid, d);
+
+      cluster_sizes_view = raft::make_device_vector_view<const idx_t, idx_t>(clusterSizes_ptr, mid);
+
+      params.n_clusters = mid;
+
+      raft::cluster::detail::kmeans_fit_predict<value_t, idx_t>(handle,
+                                                                params,
+                                                                X,
+                                                                std::nullopt,
+                                                                std::make_optional(centroids_view),
+                                                                labels.view(),
+                                                                residual,
+                                                                n_iter);
+
+      detail::countLabels(
+        handle, labels.data_handle(), clusterSizes.data_handle(), n, mid, workspace);
+
+      resultsView[mid]           = residual[0];
+      clusterDispertionView[mid] = raft::stats::cluster_dispersion(
+        handle, centroids_const_view, cluster_sizes_view, std::nullopt, n);
```

Comment:
This block of code appears 3x in this function  (here, and for calculating left/right values above)- maybe we should refactor to remove the duplicate code? I think we could pull out to a lambda fairly easily, and then call that

like 
```c++
auto calculate_dispersion = [&](int idx) { 
... 
};
calculate_dispersion(mid);
```

Reply:
Good point! I'll fix this.